server/leaderboard/calculate.js

```diff
@@ -1,70 +1,139 @@
+    }
   }
-}
 
-for (let i = 0; i < allChallenges.length; i++) {
-  const challenge = allChallenges[i]
-  if (!solveAmount.has(challenge.id)) {
-    // There are currently no solves
-    challengeValues.set(challenge.id, utilScores.getScore('dynamic', challenge.points.min, challenge.points.max, 0))
-  } else {
-    challengeValues.set(challenge.id, utilScores.getScore('dynamic', challenge.points.min, challenge.points.max, solveAmount.get(challenge.id)))
+  for (let i = 0; i < users.length; i++) {
+    const user = users[i]
+    let currScore = 0
+    const solvedChalls = userSolves.get(user.id)
+    if (solvedChalls === undefined) continue // If the user has not solved any challenges, do not add to leaderboard
+    for (let j = 0; j < solvedChalls.length; j++) {
+      // Add the score for the specific solve loaded from the challengeValues array using ids
+      const value = challengeValues.get(solvedChalls[j])
+      if (value !== undefined) {
+        currScore += value
+      }
+    }
+    userScores.push([user.id, user.name, parseInt(user.division), currScore, userLastSolves.get(user.id)])
   }
-}
 
-for (let i = 0; i < users.length; i++) {
-  const user = users[i]
-  let currScore = 0
-  const solvedChalls = userSolves.get(user.id)
-  if (solvedChalls === undefined) continue // If the user has not solved any challenges, do not add to leaderboard
-  for (let j = 0; j < solvedChalls.length; j++) {
-    // Add the score for the specific solve loaded from the challengeValues array using ids
-    currScore += challengeValues.get(solvedChalls[j])
+  return {
+    challengeValues,
+    userScores
   }
-  userScores.push([user.id, user.name, parseInt(user.division), currScore, userLastSolves.get(user.id)])
 }
 
-const sortedUsers = userScores.sort((a, b) => {
+const userCompare = (a, b) => {
   // sort the users by score
   // if two user's scores are the same, sort by last solve time
   const scoreCompare = b[3] - a[3]
   if (scoreCompare !== 0) {
     return scoreCompare
   }
   return a[4] - b[4]
-}).map((user) => user.slice(0, 4))
+}
+
+if (graph) {
+  const samples = calcSamples({ start, end })
+  const output = []
+
+  samples.forEach((sample) => {
+    const { userScores } = calculateScores(sample)
+    output.push({
+      sample,
+      scores: userScores.map((score) => [score[0], score[3]])
+    })
+  })
+
+  parentPort.postMessage({
+    leaderboards: output
+  })
+} else {
+  let sample
+  let isSample = false
+  const prevSample = getPreviousSample()
+  if (lastUpdate !== prevSample) {
+    sample = prevSample
+    isSample = true
+  } else {
+    sample = Date.now()
+  }
 
-const challengeScores = []
-challengeValues.forEach((score, id) => [
-  challengeScores.push(id, score)
-])
+  const { challengeValues, userScores } = calculateScores(sample)
 
-parentPort.postMessage({
-  leaderboard: sortedUsers,
-  challengeScores
-})
+  const sortedUsers = userScores.sort(userCompare).map((user) => user.slice(0, 4))
```

Comment:
We might want a more formal definition for the `user` object, is it possible to use Typescript for this?

Reply:
We can do this in another PR.